ticTacToe.ts

```
@@ -5,84 +5,150 @@ export type CellType = "" | "X" | "O";
+    for (let row = 0; row < gameBoard.length; row++) {
+      for (let col = 0; col < gameBoard[row].length; col++) {
+        if (gameBoard[row][col]) {
+          this.countMoves++;
+
+          gameBoard[row][col] === "X"
+            ? countPlayerXMoves++
+            : countPlayerYMoves++;
+        }
+      }
+    }
+    if (
+      !this.isValidGameBoardProvided(
+        countPlayerXMoves,
+        countPlayerYMoves,
+        this.countMoves
+      )
+    ) {
+      throw new Error(Errors.invalidGameBoardInitalProvided);
+    }
+    this.curentPlayer = this.countMoves % 2 === 0 ? "X" : "O";
+  }
+
+  private isValidGameBoardProvided(
+    countPlayerXMoves: number,
+    countPlayersYMoves: number,
+    countAllMoves: number
+  ) {
+    if (countAllMoves % 2 === 0) {
+      //they should have the same amount of moves
+      return countPlayerXMoves === countPlayersYMoves;
+    } else {
+      return countPlayerXMoves - 1 === countPlayersYMoves;
+    }
+  }
+
+  public getWinner() {
     return this.winner;
   }
 
-  makeMove(row: number, col: number) {
+  private validateMove(
+    row: number,
+    col: number,
+    player: CellType
+  ): string | null {
     if (!this.checkIfValidCoordinates(row, col)) {
-      throw new Error(Errors.invalidMoveCoordinates);
+      return Errors.invalidMoveCoordinates;
     }
 
     if (!this.isCellEmpty(row, col)) {
-      throw new Error(Errors.spotAlreadyTaken);
+      return Errors.spotAlreadyTaken;
+    }
+
+    if (this.curentPlayer !== player) {
+      return Errors.otherPlayersTurnToMakeMove;
     }
+
     if (this.winner) {
-      throw new Error(Errors.cantMakeMovesAfterGameEnd);
+      return Errors.cantMakeMovesAfterGameEnd;
     }
-    this.gameBoard[row][col] = this.curentPlayer;
-    this.curentPlayer = this.curentPlayer === "X" ? "O" : "X";
-    this.winner = this.checkIfThereIsWinner();
-    this.countMoves++;
-    if (this.countMoves === 9 && !this.winner) {
-      this.winner = "None";
+
+    return null;
```

instead of returning `null` and then throwing, you can convert the function to return nothing (`void`) and throw directly from here